### ISSUE TITLE
Potential use after free in ofi_connect()

### ISSUE THREAD

**Issue by badgerious** [aws/aws-ofi-nccl#18]
The `fi_tsend()` call in `ofi_connect()` uses a local stack variable (local_ep_addr) as the send buffer. This buffer must remain valid until the completion for this send has been reaped, however, the function exits without ensuring this. This could result in a corrupted connect message if the memory is reused and overwritten before the provider has sent the data. 

Either an inject-style send should be used, or the local_ep_addr needs to live on the heap somewhere.

**Comment by rashikakheria** [comment]
Do you have a case where you have reproduced this?

**Comment by badgerious** [comment]
Yes, but I'm mainly testing on a not-yet-released provider. I tried to reproduce on tcp;ofi_rxm, but I kept running into unrelated problems.

In general, running `nccl_connection` in a loop while binding to a single core per node (so that any auto progress threads must fight with the main thread for the same CPU, and thus make slower progress) tends to produce the problem more readily. Using manual progress would likely make reproducing much easier as well.
